File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/LabeledControl.java
package com.hello.uxml.tools.framework;

import java.util.EnumSet;

/**
 * Implements a control pattern that has content and a text and image used to
 * label(adorn) the control.
 *
 * @author ferhat
 */
public class LabeledControl extends Control {

  /** Content Property Definition */
  public static PropertyDefinition contentPropDef = PropertySystem.register("Content",
      Object.class, LabeledControl.class, new PropertyData(null,
          EnumSet.of(PropertyFlags.Resize)));

  /** Label Property Definition */
  public static PropertyDefinition labelPropDef = PropertySystem.register("Label",
      String.class, LabeledControl.class, new PropertyData(null,
          EnumSet.of(PropertyFlags.Resize)));

  /** Picture Property Definition */
  public static PropertyDefinition picturePropDef = PropertySystem.register("Picture",
      UIElement.class, LabeledControl.class, new PropertyData(null,
          EnumSet.of(PropertyFlags.Resize)));

  /**
   * Sets content of container.
   */
  @ContentNode
  public void setContent(Object content) {
    setProperty(contentPropDef, content);
  }

  /**
   * Returns content.
   */
  public Object getContent() {
    return getProperty(contentPropDef);
  }

  /**
   * Sets or returns label of control.
   */
  public void setLabel(String label) {
    setProperty(labelPropDef, label);
  }

  public String getLabel() {
    return (String) getProperty(labelPropDef);
  }

  /**
   * Sets or returns picture label of control.
   */
  public void setPicture(UIElement picture) {
    setProperty(picturePropDef, picture);
  }

  public UIElement getPicture() {
    return (UIElement) getProperty(picturePropDef);
  }
}

File: framework/lib/labeled_control.dart
part of uxml;

/**
 * Implements container that decorates an inner element.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class LabeledControl extends Control {

  static ElementDef labeledcontrolElementDef;
  /** Content property definition */
  static PropertyDefinition contentProperty;
  /** Label property definition */
  static PropertyDefinition labelProperty;
  /** Content property definition */
  static PropertyDefinition pictureProperty;

  LabeledControl() : super() {
  }

  /** Sets or returns content element. */
  set content(Object content) {
    setProperty(contentProperty, content);
  }

  Object get content => getProperty(contentProperty);

  /** Sets or returns label. */
  set label(Object element) {
    setProperty(labelProperty, element);
  }

  Object get label => getProperty(labelProperty);

  /** Sets or returns picture element. */
  set picture(UIElement element) {
    setProperty(pictureProperty, element);
  }

  UIElement get picture => getProperty(pictureProperty);

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => labeledcontrolElementDef;

  /** Registers component. */
  static void registerLabeledControl() {
    labelProperty = ElementRegistry.registerProperty("label",
        PropertyType.OBJECT, PropertyFlags.NONE, null, null);
    contentProperty = ElementRegistry.registerProperty("content",
        PropertyType.OBJECT, PropertyFlags.NONE, null, null);
    pictureProperty = ElementRegistry.registerProperty("picture",
        PropertyType.UIELEMENT, PropertyFlags.NONE, null, null);
    labeledcontrolElementDef = ElementRegistry.register("LabeledControl",
        Control.controlElementDef, [contentProperty, labelProperty,
        pictureProperty], null);
  }
}
